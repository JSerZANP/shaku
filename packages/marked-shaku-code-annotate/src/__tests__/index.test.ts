import { expect, test } from "vitest";
import { Marked } from "marked";
import markedShakuCodeAnnotate from "../index";

const marked = new Marked(markedShakuCodeAnnotate());
const markedWithTheme = new Marked(
  markedShakuCodeAnnotate({
    themes: ["github-light", "github-dark"],
  })
);

test("should not apply annotation without meta", async () => {
  const result = await marked.parse(`
\`\`\`js 
const a = 1;
//    ^
//    [This is line 1]
//    [This is line two]
\`\`\`
`);

  expect(result).toMatchSnapshot();
});

test("able to render callout: ^", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate
const a = 1;
//    ^
//    [This is line 1]
//    [This is line two]
\`\`\`
`);

  const result2 = await marked.parse(`
\`\`\`js annotate
const a = 1;
/*    ^ 
      [This is line 1] 
      [This is line two] */
\`\`\`
`);

  const result3 = await marked.parse(`
\`\`\`js annotate
const a = 1;
/*    ^  */
  //  [This is line 1] 
   /* [This is line two] */
\`\`\`
`);

  const result4 = await marked.parse(`
\`\`\`js annotate
const a = 1;
/*      ^<<*/
  //    [This is line 1] <<
   /*  [This is line two]< */
\`\`\`
`);
  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
  expect(result3).toEqual(result1);
  expect(result4).toEqual(result1);
});

test("able to render solid underline: ----", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate
const a = 1;
//    -----
//    [This is line 1]
//    [This is line two]
\`\`\`
`);
  const result2 = await marked.parse(`
\`\`\`js annotate
const a = 1;
   /* -----   */
  /*  [This is line 1] 
      [This is line two] */
\`\`\`
`);
  const result3 = await marked.parse(`
\`\`\`js annotate
const a = 1;
   /*  -----<   */
  /*    [This is line 1]  <<
          [This is line two] <<<< */
\`\`\`
`);
  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
  expect(result3).toEqual(result1);
});

test("able to render dotted underline: .....", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate
const a = 1;
//    .....
//    [This is line 1]
//    [This is line two]
\`\`\`
`);
  const result2 = await marked.parse(`
\`\`\`js annotate
const a = 1;
   /* ..... */
  /*  [This is line 1]
      [This is line two] */
\`\`\`
`);
  const result3 = await marked.parse(`
\`\`\`js annotate
const a = 1;
   /*  .....< */
  /*    [This is line 1]  <<
         [This is line two]<<< */
\`\`\`
`);
  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
  expect(result3).toEqual(result1);
});

test("able to render wavy underline: ~~~~~ ", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate
const a = 1;
//    ~~~~~
//    [This is line 1]
//    [This is line two]
\`\`\`
`);

  const result2 = await marked.parse(`
\`\`\`js annotate
const a = 1;
   /* ~~~~~ */
  /*  [This is line 1] */
   /* [This is line two] */
\`\`\`
`);

  const result3 = await marked.parse(`
\`\`\`js annotate
const a = 1;
   /*  ~~~~~< */
  /*    [This is line 1] << */
   /*    [This is line two]   <<< */
\`\`\`
`);

  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
  expect(result3).toEqual(result1);
});

test("able to render underline without comments ", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate
const a = 1;
//    -----
const b = 1;
//    .....
const c = 1;
//    ~~~~~
\`\`\`
`);

  const result2 = await marked.parse(`
\`\`\`js annotate
const a = 1;
  /*  ----- */
const b = 1;
   /* ..... */
const c = 1;
    /*~~~~~ */
\`\`\`
`);
  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
});

test("able to render highlight ", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate

// @highlight
const a = 1;

// @highlight start
const b = 1;
const c = 1;
// @highlight end

\`\`\`
`);

  const result2 = await marked.parse(`
\`\`\`js annotate

/* @highlight */
const a = 1;

/*  @highlight start */
const b = 1;
const c = 1;
/*  @highlight end */

\`\`\`
`);

  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
});

test("able to render dim ", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate

// @dim
const a = 1;

// @dim start
const b = 1;
const c = 1;
// @dim end

\`\`\`
`);

  const result2 = await marked.parse(`
\`\`\`js annotate

/* @dim */
const a = 1;

/*   @dim start */
const b = 1;
const c = 1;
   /*   @dim end */

\`\`\`
`);
  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
});

test("able to render focus ", async () => {
  const result1 = await marked.parse(`
\`\`\`js annotate

// @focus
const a = 1;

// @focus start
const b = 1;
const c = 1;
  // @focus end

\`\`\`
`);

  const result2 = await marked.parse(`
\`\`\`js annotate

/* @focus */
const a = 1;

  /* @focus start */
const b = 1;
const c = 1;
  /* @focus end */

\`\`\`
`);
  expect(result1).toMatchSnapshot();
  expect(result2).toEqual(result1);
});

test("able to render  in jsx", async () => {
  const result = await marked.parse(`
\`\`\`jsx annotate
function a() {
  //       ^
  //     [This is line 1]
    return <p>aaa
    { /*      ---       */}
       {/*   [This is line 2]       */}
       {/*   @highlight       */}
       <button>click me </button>
       {123}
    </p>
  }
\`\`\`
`);
  expect(result).toMatchSnapshot();
});

test("able to render  in tsx", async () => {
  const result = await marked.parse(`
\`\`\`tsx annotate
function a() {
  //       ^
  //     [This is line 1]
    return <p>aaa
    { /*      ---       */}
       {/*   [This is line 2]       */}
       {/*   @highlight       */}
       <button>click me </button>
       {123}
    </p>
  }
\`\`\`
`);
  expect(result).toMatchSnapshot();
});

test("able to render multiple code blocks with themes", async () => {
  const result = await markedWithTheme.parse(`
\`\`\`js annotate
const a = 1;
//    ^
//    [This is line 1]
//    [This is line two]
\`\`\`
`);

  expect(result).toMatchSnapshot();
});
