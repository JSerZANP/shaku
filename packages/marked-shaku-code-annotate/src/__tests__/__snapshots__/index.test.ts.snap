// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`able to render  in jsx 1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">a</span><span style=\\"color: #24292E\\">() {</span></div><div class=\\"shaku-callout\\" style=\\"left:9ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:2ch\\"></span><p>This is line 1</p></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    </span><span style=\\"color: #D73A49\\">return</span><span style=\\"color: #24292E\\"> &lt;</span><span style=\\"color: #22863A\\">p</span><span style=\\"color: #24292E\\">&gt;aaa</span></div><div class=\\"shaku-underline shaku-underline-solid\\" style=\\"left:13ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:1ch\\">---</span><p>This is line 2</p></div><div class=\\"line highlight\\"><span style=\\"color: #24292E\\">       &lt;</span><span style=\\"color: #22863A\\">button</span><span style=\\"color: #24292E\\">&gt;click me &lt;/</span><span style=\\"color: #22863A\\">button</span><span style=\\"color: #24292E\\">&gt;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">       {</span><span style=\\"color: #005CC5\\">123</span><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    &lt;/</span><span style=\\"color: #22863A\\">p</span><span style=\\"color: #24292E\\">&gt;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  }</span></div></code></div></pre>"`;

exports[`able to render  in tsx 1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">a</span><span style=\\"color: #24292E\\">() {</span></div><div class=\\"shaku-callout\\" style=\\"left:9ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:2ch\\"></span><p>This is line 1</p></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    </span><span style=\\"color: #D73A49\\">return</span><span style=\\"color: #24292E\\"> &lt;</span><span style=\\"color: #22863A\\">p</span><span style=\\"color: #24292E\\">&gt;aaa</span></div><div class=\\"shaku-underline shaku-underline-solid\\" style=\\"left:13ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:1ch\\">---</span><p>This is line 2</p></div><div class=\\"line highlight\\"><span style=\\"color: #24292E\\">       &lt;</span><span style=\\"color: #22863A\\">button</span><span style=\\"color: #24292E\\">&gt;click me &lt;/</span><span style=\\"color: #22863A\\">button</span><span style=\\"color: #24292E\\">&gt;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">       {</span><span style=\\"color: #005CC5\\">123</span><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    &lt;/</span><span style=\\"color: #22863A\\">p</span><span style=\\"color: #24292E\\">&gt;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  }</span></div></code></div></pre>"`;

exports[`able to render callout: ^ 1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-callout\\" style=\\"left:6ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:0ch\\"></span><p>This is line 1</p><p>This is line two</p></div></code></div></pre>"`;

exports[`able to render dim  1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line dim\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"></div><div class=\\"line dim\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">b</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line dim\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">c</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"></div></code></div></pre>"`;

exports[`able to render dotted underline: ..... 1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-underline shaku-underline-dotted\\" style=\\"left:6ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:0ch\\">.....</span><p>This is line 1</p><p>This is line two</p></div></code></div></pre>"`;

exports[`able to render focus  1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line dim\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line dim\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">b</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">c</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line dim\\"></div></code></div></pre>"`;

exports[`able to render highlight  1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line highlight\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"></div><div class=\\"line highlight\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">b</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line highlight\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">c</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"></div></code></div></pre>"`;

exports[`able to render multiple code blocks with themes 1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-callout\\" style=\\"left:6ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:0ch\\"></span><p>This is line 1</p><p>This is line two</p></div></code></div></pre><pre class=\\"shiki github-dark\\" style=\\"color:#e1e4e8;background-color:#24292e\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #F97583\\">const</span><span style=\\"color: #E1E4E8\\"> </span><span style=\\"color: #79B8FF\\">a</span><span style=\\"color: #E1E4E8\\"> </span><span style=\\"color: #F97583\\">=</span><span style=\\"color: #E1E4E8\\"> </span><span style=\\"color: #79B8FF\\">1</span><span style=\\"color: #E1E4E8\\">;</span></div><div class=\\"shaku-callout\\" style=\\"left:6ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:0ch\\"></span><p>This is line 1</p><p>This is line two</p></div></code></div></pre>"`;

exports[`able to render solid underline: ---- 1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-underline shaku-underline-solid\\" style=\\"left:6ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:0ch\\">-----</span><p>This is line 1</p><p>This is line two</p></div></code></div></pre>"`;

exports[`able to render underline without comments  1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-underline shaku-underline-solid\\" style=\\"left:6ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:0ch\\">-----</span></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">b</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-underline shaku-underline-dotted\\" style=\\"left:6ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:0ch\\">.....</span></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">c</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-underline shaku-underline-wavy\\" style=\\"left:6ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:0ch\\">~~~~~</span></div></code></div></pre>"`;

exports[`able to render wavy underline: ~~~~~  1`] = `"<pre class=\\"shiki github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></div><div class=\\"shaku-underline shaku-underline-wavy\\" style=\\"left:6ch\\"><span class=\\"shaku-underline-line\\" style=\\"left:0ch\\">~~~~~</span><p>This is line 1</p><p>This is line two</p></div></code></div></pre>"`;

exports[`should be able to prevent fallback to shiki 1`] = `
"const a = 1;
//    ^
//    [This is line 1]
//    [This is line two]"
`;

exports[`should fallback to shiki if without meta 1`] = `
"<pre class=\\"shiki github-light\\" style=\\"background-color: #fff\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">a</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">1</span><span style=\\"color: #24292E\\">;</span></span>
<span class=\\"line\\"><span style=\\"color: #6A737D\\">//    ^</span></span>
<span class=\\"line\\"><span style=\\"color: #6A737D\\">//    [This is line 1]</span></span>
<span class=\\"line\\"><span style=\\"color: #6A737D\\">//    [This is line two]</span></span></code></pre>"
`;
